nidmfsl/fsl_exporter/fsl_exporter.py

```diff
@@ -428,29 +429,130 @@ def _find_inferences(self):
+                if connectivity == 26:
+                    structure = np.array([[[1, 1, 1],
+                                           [1, 1, 1],
+                                           [1, 1, 1]],
+                                          [[1, 1, 1],
+                                           [1, 1, 1],
+                                           [1, 1, 1]],
+                                          [[1, 1, 1],
+                                           [1, 1, 1],
+                                           [1, 1, 1]]], dtype='uint8')
+                else:
+                    raise Exception('Unknown connectivity: ' +
+                                    str(connectivity))
+
+                # Compute connected clusters from excursion set
+                labels, num_feat = scipy.ndimage.label(excset_img.get_data(),
+                                                       structure)
+
+                # Update labels to match FSL's table
+                # If clusters are available in voxel space
+                cluster_vox_file = glob.glob(
+                    os.path.join(analysis_dir,
+                                 'cluster*' + str(stat_num) + '.txt'))
+                if not cluster_vox_file:
+                    cluster_vox_tab = None
+                elif len(cluster_vox_file) > 1:
+                    print(cluster_vox_file)
+                    warnings.warn("Found more than 1 cluster vox file")
+                else:
+                    with warnings.catch_warnings():
+                        # Ignore "Empty input file" for no significant cluster
+                        warnings.simplefilter("ignore")
+                        cluster_vox_tab = np.loadtxt(cluster_vox_file[0],
+                                                     skiprows=1)
+
+                # If cluster vox table was not found look for coordinates in
+                # world space and convert to voxel space
+                if cluster_vox_tab is None:
+                    cluster_file = glob.glob(
+                        os.path.join(analysis_dir,
+                                     'cluster*' + str(stat_num) + '*_std.txt'))
+                    if not cluster_file:
+                        cluster_mm_tab = None
+                    elif len(cluster_file) > 1:
+                        print(cluster_file)
+                        warnings.warn("Found more than 1 cluster file")
+                    else:
+                        with warnings.catch_warnings():
+                            # Ignore "Empty input file" for no significant
+                            # cluster
+                            warnings.simplefilter("ignore")
+                            cluster_mm_tab = np.loadtxt(cluster_file[0],
+                                                        skiprows=1)
+
+                    if cluster_mm_tab is not None:
+                        # Transform cluster positions in mm into voxels
```

Comment:
The only suggestion I have for this section is many a bit more commenting explaining the stages involved in this transformation? It did take us a while to work out how to do this - perhaps it could take someone else in the future a long time to understand what is happening here.

Reply:
Ok, I've made the code slightly shorter so hopefully it reads better. The code now directly uses the `apply_affine` function from nibabel to do the transformation so not sure what we could add... What would you recommend?

Reply:
Ahhh sorry only just seen this comment. Um maybe something like:

```
# Read in coordinates of clusters in mm space.
cluster_mm = cluster_mm_tab[:, 5:8]

# Read in excursion set image header to obtain
# world to voxel mapping
excset_img = nib.load(filename)
voxToWorld = excset_img.affine
worldToVox = npla.inv(voxToWorld)

# Transform cluster coordinates to voxel space.
cluster_vox = apply_affine(worldToVox, cluster_mm)

# Record coordinates
cluster_vox_tab = cluster_mm_tab
cluster_vox_tab[:, 5:8] = cluster_vox
```

Also I've just noticed do you use `voxToWorld`?

Reply:
Thanks! Comments added in e27580c2719fff53bbc4e438344da9b259498f85. Variable `voxToWorld` was removed in 2c2a52504383de591bc070b3a069de8ff4819b7c.

Reply:
No worries! Sorry for the delayed response!